### PR TITLE
add mc ping command

### PR DESCRIPTION
**Pull request by sinhaashish**
The ouput is shown below.
```
ashish@ashish:~/code/go/src/github/minio/mc$ ./mc ping myminio
●  help.com:9000 => 284 μs
●  help.com:9001 => 235 μs
●  help.com:9002 => 234 μs
●  help.com:9003 => 224 μs

●  help.com:9000 => 490 μs
●  help.com:9001 => 364 μs
●  help.com:9002 => 360 μs
●  help.com:9003 => 256 μs

●  help.com:9000 => 809 μs
●  help.com:9001 => 781 μs
●  help.com:9002 => Error: Get "http://127.0.0.1:9002/minio/health/live": dial tcp 127.0.0.1:9002: connect: connection refused
●  help.com:9003 => 555 μs

●  help.com:9000 => 373 μs
●  help.com:9001 => 148 μs
●  help.com:9002 => Error: Get "http://127.0.0.1:9002/minio/health/live": dial tcp 127.0.0.1:9002: connect: connection refused
●  help.com:9003 => 142 μs
 ```

**In case the server for which alias is configured is down , `ping` command  continuously retries until the server is up.**
```
ashish@ashish:~/code/go/src/github/minio/mc$ ./mc ping myminio
●  http://192.168.1.9:9003 => Error: Get "http://192.168.1.9:9003/minio/admin/v3/info": dial tcp 192.168.1.9:9003: connect: connection refused

●  http://192.168.1.9:9003 => Error: Get "http://192.168.1.9:9003/minio/admin/v3/info": dial tcp 192.168.1.9:9003: connect: connection refused

●  http://192.168.1.9:9003 => Error: Get "http://192.168.1.9:9003/minio/admin/v3/info": dial tcp 192.168.1.9:9003: connect: connection refused

●  http://192.168.1.9:9003 => Error: Get "http://192.168.1.9:9003/minio/admin/v3/info": dial tcp 192.168.1.9:9003: connect: connection refused

●  help.com:9000 => 774 μs
●  help.com:9001 => 394 μs
●  help.com:9002 => 601 μs
●  help.com:9003 => 415 μs

●  help.com:9000 => 486 μs
●  help.com:9001 => 445 μs
●  help.com:9002 => 540 μs
●  help.com:9003 => 487 μs

●  help.com:9000 => 878 μs
●  help.com:9001 => 660 μs
●  help.com:9002 => 663 μs
●  help.com:9003 => 542 μs

●  help.com:9000 => 823 μs
●  help.com:9001 => 574 μs
●  help.com:9002 => 640 μs
●  help.com:9003 => 586 μs
```
**In case the domain is not set the ip:port is displayed**
```
ashish@ashish:~/code/go/src/github/minio/mc$ ./mc ping myminio
●  127.0.0.1:9000 => 356 μs
●  127.0.0.1:9001 => 248 μs
●  127.0.0.1:9002 => 227 μs
●  127.0.0.1:9003 => 237 μs

●  127.0.0.1:9000 => 898 μs
●  127.0.0.1:9001 => 719 μs
●  127.0.0.1:9002 => 698 μs
●  127.0.0.1:9003 => 615 μs

●  127.0.0.1:9000 => 776 μs
●  127.0.0.1:9001 => 782 μs
●  127.0.0.1:9002 => 748 μs
●  127.0.0.1:9003 => 792 μs

●  127.0.0.1:9000 => 939 μs
●  127.0.0.1:9001 => 849 μs
●  127.0.0.1:9002 => 675 μs
●  127.0.0.1:9003 => 679 μs



```

### CODE REVIEW

File: cmd/ping.go [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+					return fmt.Sprint(microSeconds, " micro seconds")
+				}
+				return fmt.Sprint(miliSeconds, " mili seconds")
+			}
+			return fmt.Sprint(seconds, " seconds")
+		}
+		return fmt.Sprint(minutes, " minutes")
+	}
+	return fmt.Sprint(hours, " hours")
+}
```
Comment (klauspost) [inline cmd/ping.go:114]:
```suggestion
func computeLatency(start time.Time) string {
	diff := time.Since(start)
	hours := int(diff.Hours())
	minutes := int(diff.Minutes()) % 60
	seconds := int(diff.Seconds()) % 60
	milliSeconds := int(diff.Milliseconds())
	microSeconds := int(diff.Microseconds())
	nanoSeconds := int(diff.Nanoseconds())
	switch {
	case hours > 0:
		return english.Plural(hours, "hour", "")
	case minutes > 0:
		return english.Plural(minutes, " minute", "")
	case seconds > 0:
		return english.Plural(seconds, " second", "")
	case milliSeconds > 0:
		return english.Plural(milliSeconds, " millisecond", "")
	case microSeconds > 0:
		return english.Plural(microSeconds, " microsecond", "")
	default:
		return english.Plural(nanoSeconds, " nanosecond", "")
	}
}
```

**Review by klauspost — COMMENTED** [review]
You should have a mode where it will keep trying until it gets a success, maybe where "count" acts as a maximum.

The goal of that would be to have mc wait until a cluster is up.

File: cmd/ping.go [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+		if ping.Error == "" {
+			coloredDot := console.Colorize("Info", dot)
+			// Print server title
+			msg += fmt.Sprintf("%s  %s\n", coloredDot, console.Colorize("PrintB", ping.Server))
+			msg += fmt.Sprintf("   Status: %s\n", console.Colorize("Info", ping.Status))
+			msg += fmt.Sprintf("   Latency: %s\n", ping.Latency)
+		} else {
+			coloredDot := console.Colorize("InfoFail", dot)
+			msg += fmt.Sprintf("%s  %s\n", coloredDot, console.Colorize("PrintB", ping.Server))
+			msg += fmt.Sprintf("   Status: %s\n", console.Colorize("InfoFail", ping.Status))
```
Comment (harshavardhana) [inline cmd/ping.go:136]: We should print the error as well @sinhaashish

File: cmd/ping.go [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+	// Fetch info of all servers (cluster or single server)
+	admInfo, _ := client.ServerInfo(globalContext)
+
+	httpClient := httpClient(10 * time.Second)
+	for _, server := range admInfo.Servers {
+		req, e := http.NewRequest(http.MethodGet, protocol+server.Endpoint+livelinessEndPoint, nil)
+		if e != nil {
+			return e
+		}
+		now := time.Now().UTC()
```
Comment (harshavardhana) [inline cmd/ping.go:166]: No need to have UTC for latency calculation

File: cmd/ping.go [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+	// Create a new MinIO Admin Client
+	client, err := newAdminClient(aliasedURL)
+	fatalIf(err, "Unable to initialize admin connection.")
+
+	// Fetch info of all servers (cluster or single server)
+	admInfo, _ := client.ServerInfo(globalContext)
+
+	httpClient := httpClient(10 * time.Second)
+	for _, server := range admInfo.Servers {
+		req, e := http.NewRequest(http.MethodGet, protocol+server.Endpoint+livelinessEndPoint, nil)
```
Comment (harshavardhana) [inline cmd/ping.go:162]: protocol should be picked up from the alias @sinhaashish

File: cmd/ping.go [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+			return e
+		}
+		now := time.Now().UTC()
+		resp, e := httpClient.Do(req)
+
+		latency := computeLatency(now)
+		ping.Server = server.Endpoint
+
+		if e != nil {
+			ping.Status = "error"
```
Comment (harshavardhana) [inline cmd/ping.go:173]: Which status is true error or offline

File: cmd/ping.go [hunk 6]
```diff
@@ -0,0 +1,187 @@
...
+			ping.Status = "offline"
+			ping.Error = e.Error()
+		} else {
+			ping.Status = "online"
+			ping.Latency = latency
```
Comment (harshavardhana) [inline cmd/ping.go:178]: A non 200 Ok response is also an error @sinhaashish

**Review by harshavardhana — CHANGES_REQUESTED** [review]

**Comment by harshavardhana** [comment]
isn't `mc ping `and `mc ready` basically same thing? @Praveenrajmani @sinhaashish @kannappanr - we should pick one.

**Comment by sinhaashish** [comment]
> isn't `mc ping `and `mc ready` basically same thing? @Praveenrajmani @sinhaashish @kannappanr - we should pick one.



> isn't `mc ping `and `mc ready` basically same thing? @Praveenrajmani @sinhaashish @kannappanr - we should pick one.

mc ready does the below thing
- It uses cluster readiness endpoints which checks the quorum as well for readiness
- It will check indefinitely until the cluster is ready, and exits if the cluster is online

mc ping   hits the `/live`  endpoint

File: cmd/ping.go [hunk 7]
```diff
@@ -0,0 +1,240 @@
...
+				switch resp.StatusCode {
+				case http.StatusOK:
+					ping.Latency = latency
+					ping.Error = ""
+					defer resp.Body.Close()
+				default:
+					ping.Error = resp.Status
+				}
+			}
+			pings = append(pings, ping)
```
Comment (klauspost) [inline cmd/ping.go:234]: It would be much nicer if each failed attempt wrote the result, instead of waiting for all to complete.

Comment (sinhaashish) [reply cmd/ping.go:234]: The reason for this is clubbing the o/p of all the servers of the cluster and displaying it together.

Comment (klauspost) [reply cmd/ping.go:234]: @sinhaashish Sorry, could you rephrase that, I don't really understand it?

Comment (sinhaashish) [reply cmd/ping.go:234]: As per the code we are printing `PingResult` which is the slice of `Ping` . This way we are clubbing the o/p of all the server and displaying . 
```
●  help.com:9000 => 809 μs
●  help.com:9001 => 781 μs
●  help.com:9002 => Error: Get "http://127.0.0.1:9002/minio/health/live": dial tcp 127.0.0.1:9002: connect: connection refused
●  help.com:9003 => 555 μs
```

If we instantly ping the error without waiting for other server o/p. then the servers will not be displayed together . I will be somewhat like 
```
●  help.com:9002 => Error: Get "http://127.0.0.1:9002/minio/health/live": dial tcp 127.0.0.1:9002: connect: connection refused

●  help.com:9000 => 809 μs
●  help.com:9001 => 781 μs
●  help.com:9003 => 555 μs
``` 

I thought to give the clubbed o/p of all server , that the reason i wait for response of all the servers.

**Comment by harshavardhana** [comment]
We need to change the display format here, I will provide an example of what might be a more feasible UI for this.

**Review by harshavardhana — CHANGES_REQUESTED** [review]

**Comment by harshavardhana** [comment]
@sinhaashish after spending some time on the UI we expect the UI to be something like this 

[asciicast](https://asciinema.org/a/507078)

In fact, I have all the necessary changes for this ready in case you want it in your PR.

File: cmd/ping.go [hunk 8]
```diff
@@ -0,0 +1,342 @@
...
+
+var pingFlags = []cli.Flag{
+	cli.IntFlag{
+		Name:  "count, c",
+		Usage: "perform liveliness check for count number of times",
+		Value: 4,
+	},
+	cli.DurationFlag{
+		Name:  "interval, i",
+		Usage: "Wait interval between sending ping request",
```
Comment (harshavardhana) [inline cmd/ping.go:53]:
```suggestion
		Usage: "wait interval between each request",
```

**Review by harshavardhana — CHANGES_REQUESTED** [review]
everything else LGTM - thanks @sinhaashish

**Comment by sinhaashish** [comment]
> In fact, I have all the necessary changes for this ready in case you want it in your PR.

Thanks @harshavardhana  for sharing the code.

**Review by harshavardhana — APPROVED** [review]

**Review by klauspost — APPROVED** [review]
LGTM
